internal/provider/resource_agent.go

```diff
@@ -0,0 +1,408 @@
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/humanitec/humanitec-go-autogen"
+	"github.com/humanitec/humanitec-go-autogen/client"
+)
+
+// Ensure provider defined types fully satisfy framework interfaces
+var _ resource.Resource = &Agent{}
+var _ resource.ResourceWithImportState = &Agent{}
+
+func NewResourceAgent() resource.Resource {
+	return &Agent{}
+}
+
+// Agent defines the resource implementation.
+type Agent struct {
+	client *humanitec.Client
+	orgId  string
+}
+
+// AgentModel describes the app data model.
+type AgentModel struct {
+	ID          types.String `tfsdk:"id"`
+	Description types.String `tfsdk:"description"`
+	PublicKeys  []string     `tfsdk:"public_keys"`
+}
+
+func (*Agent) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_agent"
+}
+
+func (*Agent) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "An Agent represents an instance of the Humanitec Agent that will be used by the Platform Orchestrator to deploy into a private cluster.",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "The ID of the Agent.",
+				Required:            true,
+			},
+			"description": schema.StringAttribute{
+				MarkdownDescription: "A description to show future users. It can be empty.",
+				Optional:            true,
+				Default:             stringdefault.StaticString(""),
+				Computed:            true,
+			},
+			"public_keys": schema.ListAttribute{
+				ElementType:         types.StringType,
+				MarkdownDescription: "A list of pcks8 RSA public keys PEM encoded (as the ones produced by openssl), whose module length is greater or equal than 4096 bits.",
+				Required:            true,
+			},
+		},
+	}
+}
+
+func (a *Agent) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	resdata, ok := req.ProviderData.(*HumanitecData)
+
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *http.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+
+	a.client = resdata.Client
+	a.orgId = resdata.OrgID
+}
+
+func (a *AgentModel) updateFromContent(ctx context.Context, res *client.Agent, keys *[]client.Key) {
+	a.ID = types.StringValue(res.Id)
+	if res.Description == nil {
+		a.Description = types.StringValue("")
+	} else {
+		a.Description = types.StringValue(*res.Description)
+	}
+
+	var keysMap = fromKeyListToMap(*keys)
+	var modelKeysMap = a.getKeysMap()
+
+	// keys are sorted API side by creation date but here sorting doesn't help as all the keys are created within a secondd
+	var newModelKeys = make([]string, 0)
```

Comment:
A bit of effort is required here to keep the same order after the `apply` between the state and the keys returned by the API.